File: tasks/seeds/vcfarch-0100/files/verifier.py
#!/usr/bin/env python3
"""Protected, offline verification for the VCF migration artifact."""

from __future__ import annotations

import json
import re
import subprocess
import sys
from datetime import date
from pathlib import Path
from typing import Any
from urllib.parse import urlparse


ROOT = Path(__file__).resolve().parent


class VerificationError(Exception):
    pass


def require(condition: bool, message: str) -> None:
    if not condition:
        raise VerificationError(message)


def resolve_pointer(document: Any, pointer: str) -> Any:
    require(pointer.startswith("#/"), f"only local schema references are supported: {pointer}")
    value = document
    for token in pointer[2:].split("/"):
        token = token.replace("~1", "/").replace("~0", "~")
        require(isinstance(value, dict) and token in value, f"unresolved schema reference {pointer}")
        value = value[token]
    return value


def type_matches(value: Any, expected: str) -> bool:
    if expected == "object":
        return isinstance(value, dict)
    if expected == "array":
        return isinstance(value, list)
    if expected == "string":
        return isinstance(value, str)
    if expected == "integer":
        return isinstance(value, int) and not isinstance(value, bool)
    if expected == "number":
        return isinstance(value, (int, float)) and not isinstance(value, bool)
    if expected == "boolean":
        return isinstance(value, bool)
    if expected == "null":
        return value is None
    raise VerificationError(f"unsupported schema type {expected}")


def validate_schema(value: Any, schema: dict[str, Any], document: dict[str, Any], path: str) -> None:
    """Validate the JSON Schema/OpenAPI keywords used by the two shipped schemas."""
    if "$ref" in schema:
        validate_schema(value, resolve_pointer(document, schema["$ref"]), document, path)
        return

    if value is None and schema.get("nullable") is True:
        return

    for branch in schema.get("allOf", []):
        validate_schema(value, branch, document, path)

    if "anyOf" in schema:
        successes = 0
        for branch in schema["anyOf"]:
            try:
                validate_schema(value, branch, document, path)
                successes += 1
            except VerificationError:
                pass
        require(successes >= 1, f"{path} does not satisfy any allowed schema")

    if "oneOf" in schema:
        successes = 0
        for branch in schema["oneOf"]:
            try:
                validate_schema(value, branch, document, path)
                successes += 1
            except VerificationError:
                pass
        require(successes == 1, f"{path} must satisfy exactly one allowed schema")

    if "const" in schema:
        require(value == schema["const"], f"{path} must equal {schema['const']!r}")
    if "enum" in schema:
        require(value in schema["enum"], f"{path} is not an allowed value")

    expected_type = schema.get("type")
    if expected_type is not None:
        alternatives = expected_type if isinstance(expected_type, list) else [expected_type]
        require(any(type_matches(value, item) for item in alternatives),
                f"{path} must have type {expected_type}")

    if isinstance(value, dict):
        required = schema.get("required", [])
        missing = [name for name in required if name not in value]
        require(not missing, f"{path} is missing required properties {missing}")
        properties = schema.get("properties", {})
        for name, child in value.items():
            if name in properties:
                validate_schema(child, properties[name], document, f"{path}.{name}")
            elif schema.get("additionalProperties") is False:
                raise VerificationError(f"{path} has unexpected property {name}")
            elif isinstance(schema.get("additionalProperties"), dict):
                validate_schema(child, schema["additionalProperties"], document,
                                f"{path}.{name}")
        if "minProperties" in schema:
            require(len(value) >= schema["minProperties"], f"{path} has too few properties")
        if "maxProperties" in schema:
            require(len(value) <= schema["maxProperties"], f"{path} has too many properties")

    if isinstance(value, list):
        if "minItems" in schema:
            require(len(value) >= schema["minItems"], f"{path} has too few items")
        if "maxItems" in schema:
            require(len(value) <= schema["maxItems"], f"{path} has too many items")
        if schema.get("uniqueItems"):
            encoded = [json.dumps(item, sort_keys=True, separators=(",", ":")) for item in value]
            require(len(encoded) == len(set(encoded)), f"{path} items must be unique")
        if isinstance(schema.get("items"), dict):
            for index, child in enumerate(value):
                validate_schema(child, schema["items"], document, f"{path}[{index}]")

    if isinstance(value, str):
        if "minLength" in schema:
            require(len(value) >= schema["minLength"], f"{path} is too short")
        if "maxLength" in schema:
            require(len(value) <= schema["maxLength"], f"{path} is too long")
        if "pattern" in schema:
            require(re.search(schema["pattern"], value) is not None,
                    f"{path} does not match its schema pattern")

    if isinstance(value, (int, float)) and not isinstance(value, bool):
        if "minimum" in schema:
            require(value >= schema["minimum"], f"{path} is below its minimum")
        if "maximum" in schema:
            require(value <= schema["maximum"], f"{path} is above its maximum")


def run_client() -> Any:
    build = ROOT / ".sandbox-home" / "build"
    build.mkdir(parents=True, exist_ok=True)
    compiled = subprocess.run(
        ["javac", "-d", str(build), "MigrationPlanClient.java", "TestMain.java"],
        cwd=ROOT,
        text=True,
        capture_output=True,
    )
    require(compiled.returncode == 0, "Java compilation failed:\n" + compiled.stderr[-2000:])
    executed = subprocess.run(
        ["java", "-cp", str(build), "TestMain"],
        cwd=ROOT,
        text=True,
        capture_output=True,
        timeout=30,
    )
    require(executed.returncode == 0, "TestMain failed:\n" + executed.stderr[-2000:])
    try:
        artifact = json.loads(executed.stdout)
    except json.JSONDecodeError as error:
        raise VerificationError(f"client output is not one JSON value: {error}") from error
    return artifact


def version_matches(value: str, pattern: str) -> bool:
    expression = "^" + re.escape(pattern).replace(r"\*", ".*") + "$"
    return re.fullmatch(expression, value) is not None


def verify_research_log() -> None:
    research = (ROOT / "research-consulted.md").read_text()
    require("replace this placeholder" not in research.lower(),
            "research-consulted.md still contains its placeholder")
    consulted_dates = set(re.findall(r"\b\d{4}-\d{2}-\d{2}\b", research))
    require(consulted_dates, "research-consulted.md is missing a consultation date")
    for consulted in consulted_dates:
        try:
            date.fromisoformat(consulted)
        except ValueError as error:
            raise VerificationError(
                f"invalid research consultation date {consulted}"
            ) from error

    url_pattern = re.compile(r"https://[^\s)>|]+")
    entries: dict[str, str] = {}
    paragraphs = re.split(r"\n\s*\n", research)
    for match in url_pattern.finditer(research):
        url = match.group(0).rstrip(".,;:")
        context = next((paragraph for paragraph in paragraphs if url in paragraph), "")
        entries[url] = context

    require(len(entries) >= 3,
            "research-consulted.md must contain at least three live source URLs")
    hosts: set[str] = set()
    has_interop_or_upgrade_path = False
    has_vcf_91_guidance = False
    for url, context in entries.items():
        parsed = urlparse(url)
        host = (parsed.hostname or "").lower()
        require(parsed.scheme == "https" and "." in host and
                host != "localhost" and not host.endswith(".invalid"),
                "research sources must use plausible reachable HTTPS URLs")
        hosts.add(host)
        description = url_pattern.sub("", context)
        description = re.sub(r"[#*_|`—-]+", " ", description)
        words = re.findall(r"[A-Za-z0-9.]+", description)
        require(len(words) >= 10,
                f"research source {url} is missing its title or fact used")
        combined = context.lower()
        if (host == "interopmatrix.broadcom.com" or "interoperab" in combined or
                "upgrade path" in combined or "upgrade-path" in combined):
            has_interop_or_upgrade_path = True
        if "9.1" in combined and ("vcf" in combined or "cloud foundation" in combined):
            has_vcf_91_guidance = True

    require("compatibilityguide.broadcom.com" in hosts,
            "research is missing the published Broadcom Compatibility Guide")
    require(has_interop_or_upgrade_path,
            "research is missing Broadcom interoperability/upgrade-path material")
    require(has_vcf_91_guidance,
            "research is missing relevant Broadcom VCF 9.1 upgrade guidance")


def assert_target_sddc_values(artifact: dict[str, Any], inventory: dict[str, Any],
                              snapshot: dict[str, Any]) -> None:
    spec = artifact["targetSddcSpec"]
    identity = inventory["managementIdentity"]
    target = snapshot["targetVcfVersion"]
    require(spec.get("sddcId") == identity["sddcId"], "targetSddcSpec.sddcId changed")
    require(spec.get("workflowType") == "VCF", "existing-estate workflowType must be VCF")
    require(spec.get("version") == target, "targetSddcSpec version is not the pinned target")
    require(spec.get("vcfInstanceName") == identity["vcfInstanceName"],
            "targetSddcSpec.vcfInstanceName changed")
    dns = spec.get("dnsSpec", {})
    require(dns.get("subdomain") == identity["dnsSubdomain"] and
            dns.get("nameservers") == identity["nameservers"],
            "targetSddcSpec DNS values differ from inventory")
    require(identity["managementNetwork"] in spec.get("networkSpecs", []),
            "targetSddcSpec does not preserve the inventory management network")
    vcenter = spec.get("vcenterSpec", {})
    require(vcenter.get("vcenterHostname") == identity["vcenterHostname"],
            "targetSddcSpec vCenter hostname changed")
    require(vcenter.get("rootVcenterPassword") == identity["rootVcenterPassword"],
            "targetSddcSpec vCenter fixture credential changed")
    require(vcenter.get("sslThumbprint") == identity["sslThumbprint"],
            "targetSddcSpec vCenter thumbprint changed")
    require(vcenter.get("useExistingDeployment") is True,
            "targetSddcSpec must convert the existing vCenter")
    transitions = {item["componentType"]: item for item in snapshot["componentTransitions"]}
    require(vcenter.get("version") == transitions["VCENTER"]["targetVersion"],
            "targetSddcSpec vCenter version is not the pinned target")
    license_spec = spec.get("licenseServerSpec", {})
    require(license_spec.get("hostname") == inventory["entitlement"]["licenseServerFqdn"] and
            license_spec.get("version") == transitions["LICENSE_SERVER"]["targetVersion"] and
            license_spec.get("useExistingDeployment") is False,
            "targetSddcSpec does not deploy the mandatory license server")


def verify_plan(artifact: dict[str, Any], inventory: dict[str, Any],
                snapshot: dict[str, Any]) -> None:
    require(artifact["estateId"] == inventory["estateId"], "estateId changed")
    require(artifact["sourceVcfVersion"] == inventory["currentVcfVersion"],
            "source VCF version changed")
    require(artifact["targetVcfVersion"] == inventory["requestedTargetVcfVersion"],
            "requested target VCF version changed")
    require(artifact["targetVcfVersion"] == snapshot["targetVcfVersion"],
            "target is not the pinned compatibility target")
    supported_edge = any(
        edge == {
            "from": artifact["sourceVcfVersion"],
            "to": artifact["targetVcfVersion"],
            "direct": True,
        }
        for edge in snapshot["supportedVcfUpgradeEdges"]
    )
    require(supported_edge, "the plan does not use a supported direct VCF edge")

    entitlement = inventory["entitlement"]["entitledCores"]
    eligible = [
        option for option in inventory["topologyOptions"]
        if option["supportedByTarget"]
        and option["evacuationFeasible"]
        and option["physicalCores"] <= entitlement
    ]
    require(len(eligible) == 1, "fixture does not yield one deterministic entitled topology")
    selected = eligible[0]
    architecture = artifact["architecture"]
    require(architecture["selectedTopology"] == selected["id"] and
            architecture["licensedPhysicalCores"] == selected["physicalCores"] and
            set(architecture["retainedClusters"]) == set(selected["retainClusterIds"]) and
            set(architecture["retiredClusters"]) == set(selected["retireClusterIds"]) and
            architecture["licenseServer"] == inventory["entitlement"]["licenseServerFqdn"],
            "architecture does not select the sole compatible, entitled topology")

    steps = artifact["steps"]
    require([step["order"] for step in steps] == list(range(1, len(steps) + 1)),
            "step order must be consecutive and array ordered")
    by_id = {step["componentId"]: step for step in steps}
    require(len(by_id) == len(steps), "componentId must be unique")
    components = {item["id"]: item for item in inventory["components"]}
    require(set(by_id) == set(components), "steps must cover every inventory component exactly once")
    order = {component_id: step["order"] for component_id, step in by_id.items()}
    known_gates = {gate["id"] for gate in snapshot["gateDefinitions"]}
    transitions = {item["componentType"]: item for item in snapshot["componentTransitions"]}

    required_dependencies: dict[str, set[str]] = {component_id: set() for component_id in by_id}
    required_gates: dict[str, set[str]] = {component_id: set() for component_id in by_id}
    retired_cluster_ids = set(selected["retireClusterIds"])

    for component_id, component in components.items():
        step = by_id[component_id]
        require(step["componentType"] == component["type"], f"{component_id} type changed")
        require(step["scope"] == component["scope"], f"{component_id} scope changed")
        require(step["fromVersion"] == component["currentVersion"],
                f"{component_id} current version changed")
        if component.get("clusterId") in retired_cluster_ids:
            expected_target = "DECOMMISSIONED"
            expected_action = "RETIRE"
            required_gates[component_id].update({"entitlement-core-limit", "evacuation-feasible"})
        else:
            require(component["type"] in transitions,
                    f"no pinned transition for {component['type']}")
            transition = transitions[component["type"]]
            require(version_matches(component["currentVersion"], transition["sourcePattern"]),
                    f"{component_id} source is outside its pinned transition")
            expected_target = transition["targetVersion"]
            expected_action = transition["action"]
            required_gates[component_id].add(transition["gateId"])
        require(step["targetVersion"] == expected_target,
                f"{component_id} target is not pinned")
        require(step["action"] == expected_action, f"{component_id} action is incorrect")
        require(set(step["gates"]) <= known_gates, f"{component_id} names an unknown gate")

    retired_component_ids = [
        component_id for component_id, component in components.items()
        if component.get("clusterId") in retired_cluster_ids
    ]
    require(retired_component_ids, "selected topology retires no inventory component")

    by_type: dict[str, list[str]] = {}
    for component_id, component in components.items():
        by_type.setdefault(component["type"], []).append(component_id)

    for rule in snapshot["sequenceRules"]:
        for successor in by_type.get(rule["successorType"], []):
            for predecessor in by_type.get(rule["predecessorType"], []):
                required_dependencies[successor].add(predecessor)
                required_gates[successor].add(rule["gateId"])

    for rule in snapshot["sameScopeRules"]:
        for successor in by_type.get(rule["successorType"], []):
            if by_id[successor]["action"] == "RETIRE":
                continue
            scope = components[successor]["scope"]
            for predecessor in by_type.get(rule["predecessorType"], []):
                if components[predecessor]["scope"] == scope:
                    required_dependencies[successor].add(predecessor)
                    required_gates[successor].add(rule["gateId"])

    for rule in snapshot["scopeRules"]:
        predecessors = [
            component_id for component_id, component in components.items()
            if component["scope"] == rule["predecessorScope"]
            and component["type"] in rule["componentTypes"]
            and by_id[component_id]["action"] != "RETIRE"
        ]
        successors = [
            component_id for component_id, component in components.items()
            if component["scope"] == rule["successorScope"]
            and component["type"] in rule["componentTypes"]
            and by_id[component_id]["action"] != "RETIRE"
        ]
        for successor in successors:
            required_dependencies[successor].update(predecessors)
            required_gates[successor].add(rule["gateId"])

    def dependency_ancestors(component_id: str) -> set[str]:
        ancestors: set[str] = set()
        pending = list(by_id[component_id]["dependsOn"])
        while pending:
            dependency = pending.pop()
            if dependency not in ancestors:
                ancestors.add(dependency)
                pending.extend(by_id[dependency]["dependsOn"])
        return ancestors

    for component_id, step in by_id.items():
        dependencies = set(step["dependsOn"])
        require(dependencies <= set(by_id), f"{component_id} has an unknown dependency")
        require(all(order[dependency] < order[component_id] for dependency in dependencies),
                f"{component_id} has a dependency that is not earlier")
        require(required_dependencies[component_id] <= dependency_ancestors(component_id),
                f"{component_id} does not depend on all pinned ordering predecessors")
        require(set(step["gates"]) == required_gates[component_id],
                f"{component_id} does not name exactly its applicable pinned gates")


def main() -> int:
    try:
        artifact = run_client()

        # Required first artifact check: validate the embedded target against the
        # installer specification's own SddcSpec before loading grading fixtures.
        openapi = json.loads((ROOT / "specifications" / "vcf-installer" /
                              "vcf-installer-openapi.json").read_text())
        sddc_schema = resolve_pointer(openapi, "#/components/schemas/SddcSpec")
        target_sddc = artifact.get("targetSddcSpec") if isinstance(artifact, dict) else None
        validate_schema(target_sddc, sddc_schema, openapi,
                        "$.targetSddcSpec")

        plan_schema = json.loads((ROOT / "schemas" / "migration-plan.schema.json").read_text())
        validate_schema(artifact, plan_schema, plan_schema, "$")
        inventory = json.loads((ROOT / "fixtures" / "estate-inventory.json").read_text())
        snapshot = json.loads((ROOT / "fixtures" / "compatibility-snapshot.json").read_text())
        assert_target_sddc_values(artifact, inventory, snapshot)
        verify_plan(artifact, inventory, snapshot)
        verify_research_log()
    except (OSError, subprocess.TimeoutExpired, VerificationError) as error:
        print(f"FAIL: {error}", file=sys.stderr)
        return 1
    print("PASS: official SddcSpec, migration schema, inventory, compatibility, ordering, entitlement, and research log")
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: tasks/seeds/vcfarch-0089/files/verifier.go
package vcfplan

import (
	"bytes"
	"crypto/sha256"
	"encoding/hex"
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"
	"sort"
)

const (
	installerSchemaPath = "specifications/vcf-installer/vcf-installer-openapi.json"
	planSchemaPath      = "migration-plan.schema.json"
	planPath            = "migration-plan.json"
	estatePath          = "fixtures/estate.json"
	snapshotPath        = "fixtures/compatibility-snapshot.json"
)

type StageError struct {
	Stage string
	Err   error
}

func (e *StageError) Error() string { return e.Stage + ": " + e.Err.Error() }
func (e *StageError) Unwrap() error { return e.Err }

type Estate struct {
	EstateID         string `json:"estateId"`
	Scope            string `json:"scope"`
	VCFVersion       string `json:"vcfVersion"`
	TargetVCFVersion string `json:"targetVcfVersion"`
	Components       []struct {
		ID      string `json:"id"`
		Name    string `json:"name"`
		Version string `json:"version"`
		Role    string `json:"role"`
	} `json:"components"`
	TargetSpecInputs struct {
		SDDCID            string   `json:"sddcId"`
		WorkflowType      string   `json:"workflowType"`
		VcenterHostname   string   `json:"vcenterHostname"`
		DNSSubdomain      string   `json:"dnsSubdomain"`
		NameServers       []string `json:"nameServers"`
		ManagementNetwork struct {
			NetworkType string `json:"networkType"`
			VLANID      int    `json:"vlanId"`
			Subnet      string `json:"subnet"`
			Gateway     string `json:"gateway"`
			SubnetMask  string `json:"subnetMask"`
			MTU         int    `json:"mtu"`
		} `json:"managementNetwork"`
	} `json:"targetSpecInputs"`
}

type CompatibilitySnapshot struct {
	SnapshotVersion  string `json:"snapshotVersion"`
	EstateVCFVersion string `json:"estateVcfVersion"`
	TargetVCFVersion string `json:"targetVcfVersion"`
	InstallerSchema  struct {
		Repository string `json:"repository"`
		Tag        string `json:"tag"`
		Component  string `json:"component"`
		SHA256     string `json:"sha256"`
	} `json:"installerSchema"`
	Components []SnapshotComponent `json:"components"`
}

type SnapshotComponent struct {
	ID            string   `json:"id"`
	TargetProduct string   `json:"targetProduct"`
	TargetVersion string   `json:"targetVersion"`
	Transition    string   `json:"transition"`
	OrderRank     int      `json:"orderRank"`
	RequiredGates []string `json:"requiredGates"`
	DependsOn     []string `json:"dependsOn"`
	AllowedEdges  []Edge   `json:"allowedEdges"`
}

type Edge struct {
	FromVersion   string           `json:"fromVersion"`
	ToVersion     string           `json:"toVersion"`
	Action        string           `json:"action"`
	OrderRank     int              `json:"orderRank"`
	RequiredGates []string         `json:"requiredGates"`
	DependsOn     []EdgeDependency `json:"dependsOn"`
}

type EdgeDependency struct {
	ComponentID string `json:"componentId"`
	ToVersion   string `json:"toVersion"`
}

type Plan struct {
	APIVersion string `json:"apiVersion"`
	Kind       string `json:"kind"`
	Metadata   struct {
		EstateID         string `json:"estateId"`
		SourceVCFVersion string `json:"sourceVcfVersion"`
		TargetVCFVersion string `json:"targetVcfVersion"`
	} `json:"metadata"`
	ConsultedSources json.RawMessage `json:"consultedSources,omitempty"`
	TargetSDDCSpec   map[string]any  `json:"targetSddcSpec"`
	Components       []PlanComponent `json:"components"`
	Steps            []PlanStep      `json:"steps"`
}

type PlanComponent struct {
	ID             string `json:"id"`
	Name           string `json:"name"`
	CurrentVersion string `json:"currentVersion"`
	Target         struct {
		Product    string `json:"product"`
		Version    string `json:"version"`
		Transition string `json:"transition"`
	} `json:"target"`
	Gates []string `json:"gates"`
}

type PlanStep struct {
	Order       int      `json:"order"`
	ID          string   `json:"id"`
	ComponentID string   `json:"componentId"`
	Action      string   `json:"action"`
	FromVersion string   `json:"fromVersion"`
	ToVersion   string   `json:"toVersion"`
	Gates       []string `json:"gates"`
	DependsOn   []string `json:"dependsOn"`
}

// Verify validates the submitted architecture. Artifact JSON decoding is the
// only operation before the embedded desired state is validated against the
// pinned installer specification's own SddcSpec schema.
func Verify(root string) error {
	planDocument, err := readJSONMap(filepath.Join(root, planPath))
	if err != nil {
		return &StageError{Stage: "artifact-decode", Err: err}
	}

	openAPIBytes, err := os.ReadFile(filepath.Join(root, installerSchemaPath))
	if err != nil {
		return &StageError{Stage: "installer-schema", Err: err}
	}
	openAPI, err := decodeJSONMap(openAPIBytes)
	if err != nil {
		return &StageError{Stage: "installer-schema", Err: err}
	}
	targetSpec := planDocument["targetSddcSpec"]
	if err := validateOpenAPIComponent(openAPI, "SddcSpec", targetSpec); err != nil {
		return &StageError{Stage: "installer-schema", Err: err}
	}

	planSchema, err := readJSONMap(filepath.Join(root, planSchemaPath))
	if err != nil {
		return &StageError{Stage: "plan-schema", Err: err}
	}
	if err := ValidateJSONSchema(planSchema, planSchema, planDocument); err != nil {
		return &StageError{Stage: "plan-schema", Err: err}
	}

	var plan Plan
	if err := remarshal(planDocument, &plan); err != nil {
		return &StageError{Stage: "plan-schema", Err: err}
	}
	var estate Estate
	if err := readJSON(filepath.Join(root, estatePath), &estate); err != nil {
		return &StageError{Stage: "fixture", Err: err}
	}
	var snapshot CompatibilitySnapshot
	if err := readJSON(filepath.Join(root, snapshotPath), &snapshot); err != nil {
		return &StageError{Stage: "snapshot", Err: err}
	}
	actualHash := sha256.Sum256(openAPIBytes)
	if hex.EncodeToString(actualHash[:]) != snapshot.InstallerSchema.SHA256 {
		return &StageError{Stage: "snapshot", Err: fmt.Errorf("installer schema checksum differs from pinned tag %s", snapshot.InstallerSchema.Tag)}
	}
	if err := validatePlanSemantics(plan, estate, snapshot); err != nil {
		return &StageError{Stage: "compatibility", Err: err}
	}
	return nil
}

func validateOpenAPIComponent(openAPI map[string]any, component string, value any) error {
	components, ok := openAPI["components"].(map[string]any)
	if !ok {
		return fmt.Errorf("OpenAPI document has no components object")
	}
	schemas, ok := components["schemas"].(map[string]any)
	if !ok {
		return fmt.Errorf("OpenAPI document has no components.schemas object")
	}
	schema, ok := schemas[component].(map[string]any)
	if !ok {
		return fmt.Errorf("OpenAPI component %q is absent", component)
	}
	return ValidateJSONSchema(openAPI, schema, value)
}

func validatePlanSemantics(plan Plan, estate Estate, snapshot CompatibilitySnapshot) error {
	if estate.VCFVersion != snapshot.EstateVCFVersion || estate.TargetVCFVersion != snapshot.TargetVCFVersion {
		return fmt.Errorf("estate and compatibility snapshot VCF versions disagree")
	}
	if plan.Metadata.EstateID != estate.EstateID || plan.Metadata.SourceVCFVersion != estate.VCFVersion || plan.Metadata.TargetVCFVersion != estate.TargetVCFVersion {
		return fmt.Errorf("plan metadata does not identify the fixture estate and VCF versions")
	}
	if err := validateTargetSpecInputs(plan.TargetSDDCSpec, estate, snapshot.TargetVCFVersion); err != nil {
		return err
	}

	inventory := make(map[string]struct{ Name, Version string }, len(estate.Components))
	for _, component := range estate.Components {
		if _, duplicate := inventory[component.ID]; duplicate {
			return fmt.Errorf("duplicate inventory component %q", component.ID)
		}
		inventory[component.ID] = struct{ Name, Version string }{component.Name, component.Version}
	}
	authority := make(map[string]SnapshotComponent, len(snapshot.Components))
	for _, component := range snapshot.Components {
		authority[component.ID] = component
	}
	if len(inventory) != len(authority) {
		return fmt.Errorf("inventory and snapshot component counts differ")
	}

	plannedComponents := make(map[string]PlanComponent, len(plan.Components))
	for _, component := range plan.Components {
		if _, duplicate := plannedComponents[component.ID]; duplicate {
			return fmt.Errorf("component %q is named more than once", component.ID)
		}
		current, exists := inventory[component.ID]
		if !exists {
			return fmt.Errorf("component %q is not in the estate", component.ID)
		}
		expected, exists := authority[component.ID]
		if !exists {
			return fmt.Errorf("component %q has no compatibility authority", component.ID)
		}
		if component.Name != current.Name || component.CurrentVersion != current.Version {
			return fmt.Errorf("component %q does not preserve its inventory name and version", component.ID)
		}
		if component.Target.Product != expected.TargetProduct || component.Target.Version != expected.TargetVersion || component.Target.Transition != expected.Transition {
			return fmt.Errorf("component %q has a target outside the pinned snapshot", component.ID)
		}
		if !sameStrings(component.Gates, expected.RequiredGates) {
			return fmt.Errorf("component %q gates differ from the pinned snapshot", component.ID)
		}
		plannedComponents[component.ID] = component
	}
	if len(plannedComponents) != len(inventory) {
		return fmt.Errorf("plan must name every inventory component exactly once")
	}

	expectedStepCount := 0
	for _, component := range authority {
		expectedStepCount += len(component.AllowedEdges)
	}
	if len(plan.Steps) != expectedStepCount {
		return fmt.Errorf("plan must realize every edge in the pinned component paths")
	}
	type matchedStep struct {
		step PlanStep
		edge Edge
	}
	matchedByEdge := make(map[string]matchedStep, len(plan.Steps))
	stepsByComponent := make(map[string][]PlanStep, len(inventory))
	stepIDs := make(map[string]struct{}, len(plan.Steps))
	lastRank := -1
	for index, step := range plan.Steps {
		if step.Order != index+1 {
			return fmt.Errorf("steps must have contiguous order values beginning at 1")
		}
		if _, duplicate := stepIDs[step.ID]; duplicate {
			return fmt.Errorf("step id %q is duplicated", step.ID)
		}
		expected, exists := authority[step.ComponentID]
		if !exists {
			return fmt.Errorf("step %q names unknown component %q", step.ID, step.ComponentID)
		}
		matchedEdge, exists := findAllowedEdge(expected.AllowedEdges, step)
		if !exists {
			return fmt.Errorf("step %q uses an unsupported version edge", step.ID)
		}
		key := edgeKey(step.ComponentID, matchedEdge.ToVersion)
		if _, duplicate := matchedByEdge[key]; duplicate {
			return fmt.Errorf("step %q duplicates a pinned path edge", step.ID)
		}
		if matchedEdge.OrderRank < lastRank {
			return fmt.Errorf("step %q violates the pinned component order", step.ID)
		}
		lastRank = matchedEdge.OrderRank
		if !sameStrings(step.Gates, matchedEdge.RequiredGates) {
			return fmt.Errorf("step %q gates differ from the pinned snapshot", step.ID)
		}
		matchedByEdge[key] = matchedStep{step: step, edge: matchedEdge}
		stepsByComponent[step.ComponentID] = append(stepsByComponent[step.ComponentID], step)
		stepIDs[step.ID] = struct{}{}
	}

	for componentID, expected := range authority {
		componentSteps := stepsByComponent[componentID]
		if len(componentSteps) != len(expected.AllowedEdges) {
			return fmt.Errorf("component %q does not realize its complete pinned path", componentID)
		}
		currentVersion := inventory[componentID].Version
		for _, step := range componentSteps {
			if step.FromVersion != currentVersion {
				return fmt.Errorf("component %q path is not contiguous from inventory version %q", componentID, currentVersion)
			}
			currentVersion = step.ToVersion
		}
		if currentVersion != expected.TargetVersion {
			return fmt.Errorf("component %q path does not reach target %q", componentID, expected.TargetVersion)
		}
	}

	for _, submittedStep := range plan.Steps {
		matched := matchedByEdge[edgeKey(submittedStep.ComponentID, submittedStep.ToVersion)]
		requiredStepIDs := make([]string, 0, len(matched.edge.DependsOn))
		for _, dependency := range matched.edge.DependsOn {
			predecessor, exists := matchedByEdge[edgeKey(dependency.ComponentID, dependency.ToVersion)]
			if !exists || predecessor.step.Order >= matched.step.Order {
				return fmt.Errorf("step %q does not follow pinned dependency %s@%s", matched.step.ID, dependency.ComponentID, dependency.ToVersion)
			}
			requiredStepIDs = append(requiredStepIDs, predecessor.step.ID)
		}
		if !sameStrings(matched.step.DependsOn, requiredStepIDs) {
			return fmt.Errorf("step %q dependency ids do not express its pinned gates", matched.step.ID)
		}
	}
	return nil
}

func validateTargetSpecInputs(spec map[string]any, estate Estate, targetVersion string) error {
	input := estate.TargetSpecInputs
	if spec["sddcId"] != input.SDDCID || spec["workflowType"] != input.WorkflowType || spec["version"] != targetVersion {
		return fmt.Errorf("targetSddcSpec does not retain the fixture identity, workflow, and target version")
	}
	vcenter, _ := spec["vcenterSpec"].(map[string]any)
	if vcenter["vcenterHostname"] != input.VcenterHostname || vcenter["useExistingDeployment"] != true || vcenter["version"] != targetVersion {
		return fmt.Errorf("targetSddcSpec does not retain the existing target vCenter")
	}
	dns, _ := spec["dnsSpec"].(map[string]any)
	if dns["subdomain"] != input.DNSSubdomain || !sameJSONStrings(dns["nameservers"], input.NameServers) {
		return fmt.Errorf("targetSddcSpec DNS differs from the fixture")
	}
	networks, _ := spec["networkSpecs"].([]any)
	if len(networks) != 1 {
		return fmt.Errorf("targetSddcSpec must contain the fixture management network")
	}
	network, _ := networks[0].(map[string]any)
	expected := input.ManagementNetwork
	if network["networkType"] != expected.NetworkType || !numberEquals(network["vlanId"], expected.VLANID) || network["subnet"] != expected.Subnet || network["gateway"] != expected.Gateway || network["subnetMask"] != expected.SubnetMask || !numberEquals(network["mtu"], expected.MTU) {
		return fmt.Errorf("targetSddcSpec management network differs from the fixture")
	}
	return nil
}

func findAllowedEdge(edges []Edge, step PlanStep) (Edge, bool) {
	for _, edge := range edges {
		if edge.FromVersion == step.FromVersion && edge.ToVersion == step.ToVersion && edge.Action == step.Action {
			return edge, true
		}
	}
	return Edge{}, false
}

func edgeKey(componentID, toVersion string) string {
	return componentID + "\x00" + toVersion
}

func sameStrings(left, right []string) bool {
	if len(left) != len(right) {
		return false
	}
	a, b := append([]string(nil), left...), append([]string(nil), right...)
	sort.Strings(a)
	sort.Strings(b)
	for index := range a {
		if a[index] != b[index] {
			return false
		}
	}
	return true
}

func sameJSONStrings(value any, expected []string) bool {
	raw, ok := value.([]any)
	if !ok || len(raw) != len(expected) {
		return false
	}
	actual := make([]string, len(raw))
	for index, entry := range raw {
		actual[index], ok = entry.(string)
		if !ok {
			return false
		}
	}
	return sameStrings(actual, expected)
}

func numberEquals(value any, expected int) bool {
	number, ok := schemaNumber(value)
	return ok && number == float64(expected)
}

func readJSONMap(path string) (map[string]any, error) {
	data, err := os.ReadFile(path)
	if err != nil {
		return nil, err
	}
	return decodeJSONMap(data)
}

func decodeJSONMap(data []byte) (map[string]any, error) {
	decoder := json.NewDecoder(bytes.NewReader(data))
	decoder.UseNumber()
	var document map[string]any
	if err := decoder.Decode(&document); err != nil {
		return nil, err
	}
	if decoder.More() {
		return nil, fmt.Errorf("multiple JSON documents are not allowed")
	}
	return document, nil
}

func readJSON(path string, target any) error {
	data, err := os.ReadFile(path)
	if err != nil {
		return err
	}
	decoder := json.NewDecoder(bytes.NewReader(data))
	decoder.DisallowUnknownFields()
	return decoder.Decode(target)
}

func remarshal(value any, target any) error {
	data, err := json.Marshal(value)
	if err != nil {
		return err
	}
	decoder := json.NewDecoder(bytes.NewReader(data))
	decoder.DisallowUnknownFields()
	return decoder.Decode(target)
}
